src/contour/TerminalSession.cpp

```diff
@@ -358,6 +363,28 @@ void TerminalSession::onClosed()
         terminatedAndWaitingForKeyPress_ = true;
         return;
     }
+    if (config().profile().sessionResume)
+    {
+        auto sessionFile = crispy::xdgStateHome() / "contour/session";
+        std::ofstream file(sessionFile);
+        if (!file.is_open())
+            TerminalLog()("Failed to open session file: {}", sessionFile.string());
+        auto configPath = FileSystem::absolute(config().backingFilePath).string() + "\n";
+        auto activeProfileName = profileName_ + "\n";
+        file.write(configPath.data(), configPath.size());
+        file.write(activeProfileName.data(), activeProfileName.size());
+        const auto& grid = terminal().primaryScreen().grid();
+        auto result = std::stringstream {};
+        auto writer = VTWriter(result);
+        for (int const line:
+             ranges::views::iota(-unbox<int>(grid.historyLineCount()), unbox<int>(grid.pageSize().lines)))
+        {
+            writer.write(grid.lineAt(LineOffset(line)));
+            writer.crlf();
+        }
```

Comment:
This fits very good into a public API function that can then be also unit-testable (or simply used by unit tests if fitting).
The frontend logic would then simply call this function to serialize the grid contents.